File: internal/config/merge.go
package config

import (
	"fmt"
	"os/exec"
	"strings"

	"github.com/infraflakes/sro/internal/dsl/ast"
)

// resolveBacktickLit resolves interpolation in a BacktickLit using the provided vars map.
func resolveBacktickLit(lit *ast.BacktickLit, vars map[string]string) (string, error) {
	return ResolveBacktickLitWithPos(lit, vars, 0, 0)
}

// ResolveBacktickLitWithPos resolves interpolation in a BacktickLit using the provided vars map,
// with optional line/col information for error messages.
func ResolveBacktickLitWithPos(lit *ast.BacktickLit, vars map[string]string, line, col int) (string, error) {
	var sb strings.Builder
	for _, part := range lit.Parts {
		if part.IsVar {
			val, ok := vars[part.Value]
			if !ok {
				if line > 0 || col > 0 {
					return "", fmt.Errorf("%d:%d: undefined variable ${%s}", line, col, part.Value)
				}
				return "", fmt.Errorf("undefined variable: ${%s}", part.Value)
			}
			sb.WriteString(val)
		} else {
			sb.WriteString(part.Value)
		}
	}
	return sb.String(), nil
}

func mergeShellDecl(cfg *Config, s *ast.ShellDecl) error {
	if cfg.Shell != "" {
		return fmt.Errorf("duplicate shell declaration")
	}
	cfg.Shell = s.Value
	return nil
}

func mergeSanctuaryDecl(cfg *Config, s *ast.SanctuaryDecl) error {
	if cfg.Sanctuary != "" {
		return fmt.Errorf("duplicate sanctuary declaration")
	}
	switch v := s.Value.(type) {
	case *ast.BacktickLit:
		resolved, err := resolveBacktickLit(v, cfg.Vars)
		if err != nil {
			return err
		}
		cfg.Sanctuary = resolved
	case *ast.VarRef:
		resolved, ok := cfg.Vars[v.Name]
		if !ok {
			return fmt.Errorf("undefined variable: $%s", v.Name)
		}
		cfg.Sanctuary = resolved
	}
	return nil
}

func mergeVarDecl(cfg *Config, s *ast.VarDecl) error {
	name := s.Name
	if _, exists := cfg.Vars[name]; exists {
		return fmt.Errorf("duplicate variable: %s", name)
	}
	switch v := s.Value.(type) {
	case *ast.BacktickLit:
		resolved, err := resolveBacktickLit(v, cfg.Vars)
		if err != nil {
			return err
		}
		if s.VarType == "shell" {
			if cfg.Shell == "" {
				return fmt.Errorf("shell must be declared before using shell variables")
			}
			cmd := exec.Command(cfg.Shell, "-c", resolved)
			out, err := cmd.Output()
			if err != nil {
				return fmt.Errorf("shell execution failed for var %s: %w", name, err)
			}
			cfg.Vars[name] = strings.TrimRight(string(out), "\n")
		} else {
			cfg.Vars[name] = resolved
		}
	case *ast.VarRef:
		resolved, ok := cfg.Vars[v.Name]
		if !ok {
			return fmt.Errorf("undefined variable: $%s", v.Name)
		}
		cfg.Vars[name] = resolved
	}
	return nil
}

func mergeProjectDecl(cfg *Config, s *ast.ProjectDecl) error {
	if _, exists := cfg.Projects[s.Name]; exists {
		return fmt.Errorf("duplicate project: %s", s.Name)
	}
	proj := &Project{Name: s.Name, Sync: "clone"} // default
	for _, f := range s.Fields {
		var resolved string
		var err error
		switch v := f.Value.(type) {
		case *ast.BacktickLit:
			resolved, err = resolveBacktickLit(v, cfg.Vars)
			if err != nil {
				return err
			}
		case *ast.VarRef:
			var ok bool
			resolved, ok = cfg.Vars[v.Name]
			if !ok {
				return fmt.Errorf("undefined variable: $%s", v.Name)
			}
		}
		switch f.Key {
		case "url":
			proj.URL = resolved
		case "dir":
			proj.Dir = resolved
		case "sync":
			proj.Sync = resolved
		case "use":
			proj.Use = resolved
		}
	}
	cfg.Projects[s.Name] = proj
	return nil
}

func mergeFnDecl(cfg *Config, s *ast.FnDecl) error {
	if _, exists := cfg.Functions[s.Name]; exists {
		return fmt.Errorf("duplicate function: %s", s.Name)
	}
	cfg.Functions[s.Name] = s
	return nil
}

func mergeSeqDecl(cfg *Config, s *ast.SeqDecl) error {
	if _, exists := cfg.Seqs[s.Name]; exists {
		return fmt.Errorf("duplicate seq: %s", s.Name)
	}
	cfg.Seqs[s.Name] = s
	return nil
}

func mergeParDecl(cfg *Config, s *ast.ParDecl) error {
	if _, exists := cfg.Pars[s.Name]; exists {
		return fmt.Errorf("duplicate par: %s", s.Name)
	}
	cfg.Pars[s.Name] = s
	return nil
}

func merge(programs []*ast.Program) (*Config, error) {
	cfg := &Config{
		Projects:  make(map[string]*Project),
		Functions: make(map[string]*ast.FnDecl),
		Seqs:      make(map[string]*ast.SeqDecl),
		Pars:      make(map[string]*ast.ParDecl),
		Vars:      make(map[string]string),
	}

	for _, prog := range programs {
		for _, stmt := range prog.Statements {
			switch s := stmt.(type) {
			case *ast.ShellDecl:
				if err := mergeShellDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.SanctuaryDecl:
				if err := mergeSanctuaryDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.VarDecl:
				if err := mergeVarDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.ProjectDecl:
				if err := mergeProjectDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.FnDecl:
				if err := mergeFnDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.SeqDecl:
				if err := mergeSeqDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.ParDecl:
				if err := mergeParDecl(cfg, s); err != nil {
					return nil, err
				}
			case *ast.ImportDecl:
				// already handled in parseRecursive, skip
			}
		}
	}

	return cfg, nil
}

File: src/config/merge.rs
use crate::config::error::ConfigError;
use crate::config::types::{Config, Project};
use crate::dsl::ast::{Expr, Program, Stmt, VarType};
use std::collections::HashMap;
use std::io::Read;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

pub(crate) fn merge(programs: Vec<Program>) -> Result<Config, ConfigError> {
    let mut shell = String::new();
    let mut sanctuary_expr: Option<Expr> = None;
    let mut projects: HashMap<String, Project> = HashMap::new();
    let mut global_vars: HashMap<String, String> = HashMap::new();

    for program in &programs {
        for stmt in &program.stmts {
            if let Stmt::ShellDecl { value } = stmt {
                if !shell.is_empty() {
                    return Err(ConfigError::Validation(
                        "duplicate shell declaration".to_string(),
                    ));
                }
                shell = value.clone();
            }
        }
    }

    for program in &programs {
        for stmt in &program.stmts {
            if let Stmt::VarDecl {
                name,
                value,
                var_type,
            } = stmt
            {
                if global_vars.contains_key(name) {
                    return Err(ConfigError::Validation(format!(
                        "duplicate variable: {}",
                        name
                    )));
                }

                let resolved = value
                    .resolve(&global_vars)
                    .map_err(ConfigError::Validation)?;

                let final_value = if var_type == &VarType::Shell {
                    if shell.is_empty() {
                        return Err(ConfigError::Validation(
                            "shell must be declared before using shell variables".to_string(),
                        ));
                    }
                    execute_shell(&shell, &resolved)?
                } else {
                    resolved
                };

                global_vars.insert(name.clone(), final_value);
            }
        }
    }

    for program in programs {
        for stmt in program.stmts {
            match stmt {
                Stmt::ShellDecl { .. } => {}
                Stmt::SanctuaryDecl { value } => {
                    if sanctuary_expr.is_some() {
                        return Err(ConfigError::Validation(
                            "duplicate sanctuary declaration".to_string(),
                        ));
                    }
                    sanctuary_expr = Some(value);
                }
                Stmt::ProjectDecl {
                    name, fields, body, ..
                } => {
                    if projects.contains_key(&name) {
                        return Err(ConfigError::Validation(format!(
                            "duplicate project: {}",
                            name
                        )));
                    }

                    let mut project = Project {
                        name: name.clone(),
                        url: String::new(),
                        dir: String::new(),
                        sync: "clone".to_string(),
                        use_file: None,
                        branch: String::new(),
                        vars: HashMap::new(),
                        functions: HashMap::new(),
                        seqs: HashMap::new(),
                        pars: HashMap::new(),
                    };

                    for field in &fields {
                        let value = field
                            .value
                            .resolve(&global_vars)
                            .map_err(ConfigError::Validation)?;
                        match field.key.as_str() {
                            "url" => project.url = value,
                            "dir" => project.dir = value,
                            "sync" => project.sync = value,
                            "use" => {
                                if !value.is_empty() {
                                    project.use_file = Some(value);
                                }
                            }
                            "branch" => project.branch = value,
                            _ => {
                                return Err(ConfigError::Validation(format!(
                                    "unknown project field: {}",
                                    field.key
                                )));
                            }
                        }
                    }

                    for body_stmt in body {
                        merge_project_body_stmt(&mut project, body_stmt, &shell)?;
                    }

                    projects.insert(name, project);
                }
                _ => {}
            }
        }
    }

    let sanctuary = match sanctuary_expr {
        Some(ref expr) => expr
            .resolve(&global_vars)
            .map_err(ConfigError::Validation)?,
        None => String::new(),
    };

    Ok(Config {
        shell,
        sanctuary,
        projects,
        vars: global_vars,
    })
}

pub(crate) fn merge_project_body_stmt(
    project: &mut Project,
    stmt: Stmt,
    shell: &str,
) -> Result<(), ConfigError> {
    match stmt {
        Stmt::VarDecl {
            name,
            value,
            var_type,
        } => {
            if project.vars.contains_key(&name) {
                return Err(ConfigError::Validation(format!(
                    "duplicate variable in project '{}': {}",
                    project.name, name
                )));
            }

            let resolved = value
                .resolve(&project.vars)
                .map_err(ConfigError::Validation)?;

            let final_value = if var_type == VarType::Shell {
                if shell.is_empty() {
                    return Err(ConfigError::Validation(
                        "shell must be declared before using shell variables".to_string(),
                    ));
                }
                execute_shell(shell, &resolved)?
            } else {
                resolved
            };

            project.vars.insert(name, final_value);
        }
        Stmt::FnDecl { name, body, .. } => {
            if project.functions.contains_key(&name) {
                return Err(ConfigError::Validation(format!(
                    "duplicate function in project '{}': {}",
                    project.name, name
                )));
            }
            project.functions.insert(name, body);
        }
        Stmt::SeqDecl { name, fns, .. } => {
            if project.seqs.contains_key(&name) {
                return Err(ConfigError::Validation(format!(
                    "duplicate seq in project '{}': {}",
                    project.name, name
                )));
            }
            project.seqs.insert(name, fns);
        }
        Stmt::ParDecl { name, fns, .. } => {
            if project.pars.contains_key(&name) {
                return Err(ConfigError::Validation(format!(
                    "duplicate par in project '{}': {}",
                    project.name, name
                )));
            }
            project.pars.insert(name, fns);
        }
        _ => {}
    }
    Ok(())
}

const SHELL_TIMEOUT: Duration = Duration::from_secs(30);
const POLL_INTERVAL: Duration = Duration::from_millis(50);

fn collect_output(child: &mut std::process::Child) -> (Vec<u8>, Vec<u8>) {
    let mut stdout_buf = Vec::new();
    let mut stderr_buf = Vec::new();
    let _ = child
        .stdout
        .as_mut()
        .map(|s| s.read_to_end(&mut stdout_buf));
    let _ = child
        .stderr
        .as_mut()
        .map(|s| s.read_to_end(&mut stderr_buf));
    (stdout_buf, stderr_buf)
}

pub(crate) fn execute_shell(shell: &str, command: &str) -> Result<String, ConfigError> {
    let mut child = Command::new(shell)
        .arg("-c")
        .arg(command)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| ConfigError::Validation(format!("shell execution failed: {}", e)))?;

    let start = Instant::now();
    let status = loop {
        match child.try_wait() {
            Ok(Some(status)) => break status,
            Ok(None) => {
                if start.elapsed() >= SHELL_TIMEOUT {
                    child.kill().ok();
                    child.wait().ok();
                    let (stdout_buf, stderr_buf) = collect_output(&mut child);
                    let stderr = String::from_utf8_lossy(&stderr_buf);
                    let stdout = String::from_utf8_lossy(&stdout_buf);
                    let detail = if stderr.trim().is_empty() {
                        stdout.trim().to_string()
                    } else {
                        stderr.trim().to_string()
                    };
                    return if detail.is_empty() {
                        Err(ConfigError::Validation(format!(
                            "shell command timed out after {}s",
                            SHELL_TIMEOUT.as_secs()
                        )))
                    } else {
                        Err(ConfigError::Validation(format!(
                            "shell command timed out after {}s: {}",
                            SHELL_TIMEOUT.as_secs(),
                            detail
                        )))
                    };
                }
                std::thread::sleep(POLL_INTERVAL);
            }
            Err(e) => {
                return Err(ConfigError::Validation(format!(
                    "shell execution failed: {}",
                    e
                )));
            }
        }
    };

    let (stdout_buf, stderr_buf) = collect_output(&mut child);

    if !status.success() {
        let stderr = String::from_utf8_lossy(&stderr_buf);
        return Err(ConfigError::Validation(format!(
            "shell command failed: {}",
            stderr.trim()
        )));
    }

    let stdout = String::from_utf8_lossy(&stdout_buf);
    Ok(stdout.trim_end().to_string())
}
